### DISCUSSION TITLE
Hook for packets not processed

### DISCUSSION THREAD

**Discussion by rsmmr** [zeek/zeek#1595] [Ideas]
In a number of different contexts, a desire has come up to get access to packets that Zeek doesn't process otherwise. I think we generally could pretty easily provide a plugin hook for purpose that receives the current `Packet` structure, filled in to the degree that the packet made it through the pipeline. 

The main questions is what exactly it means that Zeek "did not process" a packet. It probably should not be simply "no analyzer looked at it" because we might hit dead ends at different points in the stack (say, an usual protocol on top of IP that's not having a further analyzer to continue with).  Instead we need some user-centric notion of "this traffic is being taken care of".

I'm wondering if the following would work:

1. We track a boolean `processed` along with each `Packet`, initialized to false. 
2. As the packet makes it through through our pipeline, any packet analyzer can set it that flag to true to indicate "I'm taking care of this", meaning that at the very least there will be something in the logs about the activity. For example, TCP/IP analysis would set this as soon as we get to the point where the traffic will for sure make it into `conn.log`.
3. If at the end of a packet's life the flag is still false, we hand it over to a new plugin hook `unprocessed_packet`
4. We probably also want to record some basic stats on how many of these we see.

Does this make sense?

**Comment by timwoj** [comment]
Should this cover packets that are completely unprocessed or should it also cover ones where packets fail to be processed for one reason or another?

I think there's a couple of assumptions you could easily make on whether a packet fails to process:
- If a packet analyzer returns false. This usually generates a weird, so maybe this one doesn't matter.
- If a protocol_violation event is sent.

There's definitely other easy ones but those two came to mind right away.

**Comment by rsmmr** [comment]
The more I mull over this, the more I think it'd be quite tricky to find an "automatic" point where we consider something unprocessed, because semantics can be so different. For example, a packet analyzer returning false can mean different things (proof: the "usually" in "usually generates a weird" :-). Same for protocol violation. 

Hence my idea to use active acknowledgment to say "this was processed" and have everything else be considered "not processed".

**Reply by pbcullen** [reply]
I like the active acknowledgement. Seems similar to pkt->dump_packet which gets used after the root_analyzer->ForwardPacket() call is complete. pkt->processed could be checked at the same point in [zeek::packet_analysis::Manager::ProcessPacket()](https://github.com/zeek/zeek/blob/b372622610a4d9fb96e94d8108f282a7f6865f86/src/packet_analysis/Manager.cc#L88) then hand it over to the new unprocessed_packet hook.
